### ISSUE TITLE
Unable to handle a specific nested function-call

### ISSUE THREAD

**Issue by barakman** [ethereum/solidity#7303]
## Description

The compiler seems to be unable to "unwind" a certain function-call combination, leading to a *Stack too deep* error, which:
1. Can be resolved via a simple code-rearrangement
2. Would be resolved by "stack-oriented" compilers designated for traditional platforms

Due to each one of these reasons by itself, I believe that the Solidity compiler should be upgraded to resolve this problem as well.

## Environment

- Compiler version: 0.4.24
- Compiler optimization: enabled with runs=200
- Target EVM version: Ganache 6.5.1 if that's what you mean
- Framework/IDE: Truffle 4.1.14
- Operating system: Windows 10

## Steps to Reproduce

Consider the following contract:
```
pragma solidity 0.4.24;

contract MyContract {
    function func1(uint x1, uint x2, uint x3, uint x4, uint x5) private pure returns (uint) {
        return x1 + x2 + x3 + x4 + x5;
    }
    function func2(uint x1, uint x2, uint x3, uint x4, uint x5, uint x6, uint x7) private pure returns (uint) {
        return x1 + x2 + x3 + x4 + x5 + x6 + x7;
    }
    function func3(uint x1, uint x2, uint x3, uint x4, uint x5, uint x6, uint x7, uint x8, uint x9, uint x10) public payable returns (uint) {
        return func2(x1, x2, x3, x4, x5, x6, func1(x2, x7, x8, x9, x10));
    }
}
```
A compilation error is emitted on `func3`:
```
/C/Users/.../MyContract.sol:11:52: CompilerError: Stack too deep, try removing local variables.
        return func2(x1, x2, x3, x4, x5, x6, func1(x2, x7, x8, x9, x10));
                                                   ^^
```
In case it is not clear in the description above, those `^^` are pointing at the `x2` variable passed on the nested-call to `func1`.

This problem can easily be solved by "unwinding" the nested function-call:
```
function func3(uint x1, uint x2, uint x3, uint x4, uint x5, uint x6, uint x7, uint x8, uint x9, uint x10) public payable returns (uint) {
    uint x = func1(x2, x7, x8, x9, x10);
    return func2(x1, x2, x3, x4, x5, x6, x);
}
```
My conclusion is that the compiler should be able to do the same.
In fact, I don't quite see how the compiler ended up with this problem to begin with, as the inner function-call should be "completely unwound" before the outer function-call is dealt with. So at no point should the stack occupy more than 11 slots or so.

One point to notice is, that if I replace `func1(x2, ...)` with `func1(x6, ...)` in the original `func3`, then there is no issue with the stack. Though I'm guessing that this might be related to the fact that the nested function-call is now executed on consecutive stack-slots (`x6`, `x7`, `x8`, `x9`, `x10`), and so perhaps the compiler can optimize this specific scenario.

Thanks

**Comment by chriseth** [comment]
This will be solved by the new code generator (I actually just verified that it does). The reason why it fails currently is because function arguments are evaluated from left to right and put on the stack like that. The current code generator is too low-level to perform optimizations like that.

Sorry for the inconvenience!

**Comment by barakman** [comment]
@chriseth:
Thank you very much for your quick response.
Any chance that this fix will also be applied on the 0.4 branch (version 0.4.27 I suppose)?

**Comment by chriseth** [comment]
@barakman sorry, but that won't happen.
